[ERROR] /home/runner/work/infinispan/infinispan/core/src/test/java/org/infinispan/persistence/StoreConfigurationValidationTest.java:3:1: Import statement violating a checkstyle rule: org.testng.AssertJUnit.assertEquals. Stop using TestNG/JUnit 4.x/WildFly assertions [IllegalImport]
[ERROR] /home/runner/work/infinispan/infinispan/core/src/test/java/org/infinispan/persistence/StoreConfigurationValidationTest.java:4:1: Import statement violating a checkstyle rule: org.testng.AssertJUnit.assertTrue. Stop using TestNG/JUnit 4.x/WildFly assertions [IllegalImport]

--- a/core/src/main/java/org/infinispan/configuration/cache/PersistenceConfigurationBuilder.java
+++ b/core/src/main/java/org/infinispan/configuration/cache/PersistenceConfigurationBuilder.java
@@ -154,8 +154,13 @@ public class PersistenceConfigurationBuilder extends AbstractConfigurationChildB
             if (storeConfiguration.purgeOnStartup()) {
                throw CONFIG.sharedStoreShouldNotBePurged(storeConfiguration.getClass().getSimpleName());
             }
-         } else if (storeConfiguration.transactional() && !isLocalCache) {
-            throw CONFIG.clusteredTransactionalStoreMustBeShared(storeConfiguration.getClass().getSimpleName());
+         } else {
+            if (storeConfiguration.transactional() && !isLocalCache) {
+               throw CONFIG.clusteredTransactionalStoreMustBeShared(storeConfiguration.getClass().getSimpleName());
+            }
+            if (!isLocalCache && !storeConfiguration.purgeOnStartup()) {
+               CONFIG.nonSharedStoreWithoutPurgeOnStartup(storeConfiguration.getClass().getSimpleName());
+            }
          }
          if (storeConfiguration.async().enabled() && storeConfiguration.transactional()) {
             throw CONFIG.transactionalStoreCannotBeAsync(storeConfiguration.getClass().getSimpleName());

--- a/core/src/main/java/org/infinispan/util/logging/Log.java
+++ b/core/src/main/java/org/infinispan/util/logging/Log.java
@@ -2481,4 +2481,9 @@ public interface Log extends BasicLogger {
    @LogMessage(level = WARN)
    @Message(value = "The 'remote-site' configuration format at %s has been deprecated. Use the map-based format instead, e.g. 'NYC: { remoteSite: ~ }'", id = 727)
    void deprecatedRemoteSiteFormat(Location location);
+
+   @LogMessage(level = WARN)
+   @Message(value = "Non-shared store '%s' is configured without purgeOnStartup in a clustered cache. " +
+         "This can lead to data resurrection and inconsistency after a node restart.", id = 728)
+   void nonSharedStoreWithoutPurgeOnStartup(String storeType);
 }

--- a/core/src/test/java/org/infinispan/persistence/StoreConfigurationValidationTest.java
+++ b/core/src/test/java/org/infinispan/persistence/StoreConfigurationValidationTest.java
@@ -1,8 +1,13 @@
 package org.infinispan.persistence;
 
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
 import java.util.EnumSet;
 import java.util.Set;
 
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.core.layout.PatternLayout;
 import org.infinispan.commons.CacheConfigurationException;
 import org.infinispan.commons.configuration.BuiltBy;
 import org.infinispan.commons.configuration.ConfigurationFor;
@@ -17,6 +22,7 @@ import org.infinispan.persistence.dummy.DummyInMemoryStore;
 import org.infinispan.persistence.dummy.DummyInMemoryStoreConfiguration;
 import org.infinispan.persistence.dummy.DummyInMemoryStoreConfigurationBuilder;
 import org.infinispan.test.fwk.TestCacheManagerFactory;
+import org.infinispan.testing.skip.StringLogAppender;
 import org.testng.annotations.Test;
 
 /**
@@ -69,6 +75,45 @@ public class StoreConfigurationValidationTest {
             .addStore(DummyInMemoryStoreConfigurationBuilder.class)
             .shared(true)
             .validate();
+   }
+
+   public void testWarningNonSharedStoreWithoutPurge() {
+      Thread testThread = Thread.currentThread();
+      StringLogAppender logAppender = new StringLogAppender("org.infinispan.CONFIG",
+            Level.WARN,
+            t -> t == testThread,
+            PatternLayout.newBuilder().setPattern("%m").build());
+      logAppender.install();
+      try {
+         ConfigurationBuilder builder = TestCacheManagerFactory.getDefaultCacheConfiguration(false);
+         builder.clustering()
+               .cacheMode(CacheMode.DIST_SYNC)
+               .persistence()
+               .addStore(DummyInMemoryStoreConfigurationBuilder.class);
+         builder.build();
+         assertEquals(1, logAppender.size());
+         assertTrue(logAppender.get(0).contains("ISPN000728"));
+      } finally {
+         logAppender.uninstall();
+      }
+   }
+
+   public void testNoWarningNonSharedStoreWithoutPurgeLocalCache() {
+      Thread testThread = Thread.currentThread();
+      StringLogAppender logAppender = new StringLogAppender("org.infinispan.CONFIG",
+            Level.WARN,
+            t -> t == testThread,
+            PatternLayout.newBuilder().setPattern("%m").build());
+      logAppender.install();
+      try {
+         ConfigurationBuilder builder = TestCacheManagerFactory.getDefaultCacheConfiguration(false);
+         builder.persistence()
+               .addStore(DummyInMemoryStoreConfigurationBuilder.class);
+         builder.build();
+         assertEquals(0, logAppender.size());
+      } finally {
+         logAppender.uninstall();
+      }
    }
 
    @ConfiguredBy(NonSharedDummyStoreConfiguration.class)

